Add source generated metrics doc (#45784)
dotnet/docs: docs/core/diagnostics/metrics-generator.md

## Get started

To get started, install the [📦 Microsoft.Extensions.Telemetry.Abstractions](https://www.nuget.org/packages/Microsoft.Extensions.Telemetry.Abstractions) NuGet package:

### [.NET CLI](#tab/dotnet-cli)

```dotnetcli
dotnet add package Microsoft.Extensions.Telemetry.Abstractions
```

### [PackageReference](#tab/package-reference)

```xml
<ItemGroup>
  <PackageReference Include="Microsoft.Extensions.Telemetry.Abstractions"
                    Version="*" />
</ItemGroup>
```

---

For more information, see [dotnet add package](../tools/dotnet-package-add.md) or [Manage package dependencies in .NET applications](../tools/dependencies.md).

## Generic attributes

Generic attributes require C# 11 or later. For C# 10 or earlier, use nongeneric attributes instead.

The following example shows a class that declares two metrics. The methods are marked with an attribute and are declared as `static` and `partial`.
The code generator runs at build time and provides an implementation of these methods, along with accompanying
types.

> :::code language="csharp" source="snippets/MetricsGen/MetricConstants.cs" id="snippet_metricConstants":::

Reviewer: ```suggestion
:::code language="csharp" source="snippets/MetricsGen/MetricConstants.cs" id="constants":::
```

> :::code language="csharp" source="snippets/MetricsGen/Metrics.cs" id="snippet_Metrics" :::

Reviewer: ```suggestion
The following code demonstrates how to use the generator with primitive types:

:::code language="csharp" source="snippets/MetricsGen/Metrics.cs" id="metrics" :::
```

The previous declaration automatically returns the following:

- `Latency` class with a `Record` method
- `TotalCount` class with an `Add` method
- `TotalFailures` class with a `Add` method.

The attributes indicate the set of dimensions that each metric uses. The signature for the generated types looks like this:

```csharp
internal class TotalCount
{
    public void Add(int value, object? env, object? region, object? requestName, object? requestStatus)
}

internal TotalFailures
{
    public void Add(int value)
}

internal class Latency
{
    public void Record(long value, object? requestName, object? duration);
}
```

The dimensions specified in the attributes have been turned into arguments to the `Add` and `Record` methods. You then use the generated methods to create instances of these types. With the instances created, you can call `Add` and `Record` to register metric values, as shown in the following example:

> :::code language="csharp" source="snippets/MetricsGen/MyClass.cs" id ="snippet_metricCreation":::

Reviewer: ```suggestion
:::code language="csharp" source="snippets/MetricsGen/MyClass.cs" id ="creation":::
```
